hathor/p2p/sync_v2/transaction_streaming_client.py

```diff
@@ -84,7 +84,10 @@ def __init__(self,
         self._existing_deps: set[VertexId] = set()
 
         self._prepare_block(self.partial_blocks[0])
-        assert self._waiting_for
+
+        # XXX: it could be that we already have all dependencies so _waiting_for ends up empty, but then it would have
+        #      to be added to _existing_deps, so it must be impossible for the union of both to be empty
+        assert len(self._waiting_for) + len(self._existing_deps) > 0
```

Comment:
I think this might fix but shouldn't we try to process all blocks before starting downloading them? I mean, this fix should work but we will download transactions that we already have.

Reply:
What do you think of moving this assert to `_prepare_block` and possibly even using a builder and call `_prepare_block` on `build()` instead of on this classes `__init__` (I'd prefer if `__init__` does minimal work and avoid doing stuff like db access).

Reply:
I pushed how it would look like in a new commit, if you think it doesn't make sense I'll revert it.

Reply:
It's actually not correct to have that in `_prepare_block` it broke several tests.

Reply:
This assert was removed.

Reply:
Thanks!